Convert repository plugins to modules

pugnascotia opened the pull request:
Closes #81652.

Convert the `repository-azure`, `repository-gcs` and `repository-s3` plugins into modules, so that they are always included in the Elasticsearch distribution. Also change plugin installation, removal and syncing so that attempting to add or remove these plugins still succeeds but is now a no-op.

pugnascotia committed:
Move repository s3, gcs and azure plugins to modules

pugnascotia committed:
Update plugin-cli code for modularised plugins

pugnascotia committed:
Add test coverage for migrated plugins

pugnascotia committed:
Merge remote-tracking branch 'upstream/master' into repository-plugins-to-modules

elasticmachine commented:
Pinging @elastic/es-distributed (Team:Distributed)

elasticmachine commented:
Pinging @elastic/es-delivery (Team:Delivery)

elasticmachine commented:
Pinging @elastic/es-core-infra (Team:Core/Infra)

elasticsearchmachine commented:
Hi @pugnascotia, I've created a changelog YAML for you.

pugnascotia committed:
Update docs/changelog/81870.yaml

pugnascotia committed:
Update 81870.yaml

pugnascotia committed:
Fix test class name

pugnascotia committed:
Fix test to refer to stderr, not stdout

pugnascotia committed:
Fix test failures

pugnascotia committed:
Attempt to update docs

pugnascotia committed:
Fixy McFix-face

mark-vieira commented on distribution/tools/plugin-cli/src/main/java/org/elasticsearch/plugins/cli/InstallPluginAction.java:
```diff
@@ -219,6 +226,15 @@ public void execute(List<PluginDescriptor> plugins) throws Exception {
+                if (PLUGINS_CONVERTED_TO_MODULES.contains(pluginId)) {
+                    // This deliberately does not throw an exception in order to avoid failing automation that relies on installing this
+                    // plugin during deployment.
+                    terminal.errorPrintln(
```
I wonder if additionally sending this to stderr will trip up any folks.

mark-vieira commented on distribution/tools/plugin-cli/src/main/java/org/elasticsearch/plugins/cli/SyncPluginsAction.java:
```diff
@@ -115,7 +116,14 @@ public void execute() throws Exception {
         final List<PluginInfo> existingPlugins = getExistingPlugins();
 
-        final List<PluginDescriptor> pluginsThatShouldExist = pluginsConfig.getPlugins();
+        // For plugins that have migrated to modules, in order to help transition it's OK to still specify these plugins
```
If that's the case should we even bother filtering here? Wouldn't it be better to just delegate on and get the warning message which would inform the user to remove these from their plugin config? As it is now we'd be silencing that message, no?

pugnascotia replied:
The problem with delegating is that in this situation, the config file might say "repository-s3" should be installed, but actually we need to **remove** that plugin because it's now a module. So we need to change the "this plugin should be present" into a "this plugin should be removed". As a result, we don't reach the logic in `InstallPluginAction` that logs the warning about modularized plugins. However I realised that this implementation wasn't printing that warning at all when using a config file, so I've corrected that.

mark-vieira commented on distribution/tools/plugin-cli/src/main/java/org/elasticsearch/plugins/cli/InstallPluginAction.java:
```diff
@@ -219,6 +226,15 @@ public void execute(List<PluginDescriptor> plugins) throws Exception {
+                if (PLUGINS_CONVERTED_TO_MODULES.contains(pluginId)) {
+                    // This deliberately does not throw an exception in order to avoid failing automation that relies on installing this
+                    // plugin during deployment.
+                    terminal.errorPrintln(
```
I wonder if there are additionally any concerns with sending this to stderrr? Would stdout similarly break stuff depending how folks are consuming the output?

pugnascotia replied:
The thing is, the output from the CLI tools is completely unstructured. The only contract is the exit code of the tool, which is zero when installing or removing a modularized plugin. We can't reasonably be expected never to change the output of the CLI when it has no documented output format.

mark-vieira commented on docs/plugins/repository.asciidoc:
```diff
@@ -8,37 +8,20 @@ by distributed file systems:
-<<repository-s3,S3 Repository>>::
-
-The S3 repository plugin adds support for using S3 as a repository.
+NOTE: The S3, GCS and Azure repository plugins are now included in the {es}
```
Since they're no longer "plugins" I'd reword this to say

> NOTE: Support for S3, GCS and Azure repository is now bundled in {es} by default

I'd also rename the section "Core repository plugins" to "Official repository plugins".

@elastic/es-docs thoughts?

mark-vieira commented on modules/repository-s3/licenses/httpclient-NOTICE.txt:
```diff
@@ -1,5 +1,6 @@
 Apache HttpComponents Client
-Copyright 1999-2015 The Apache Software Foundation
+Copyright 1999-2016 The Apache Software Foundation
```
Did we update a dependency here?

pugnascotia replied:
No, but for some reason the copyright years were different between `repository-s3` and `repository-url`, so I went with the later year. The checksums of the actual jar files match.

mark-vieira reviewed (approved):
Couple of minor comments but otherwise LGTM. Looks like we have some minor packaging test fallout but I don't think there are any fundamental issues with the implementation.

pugnascotia committed:
Permit modules to contain log4j2.properties config files

pugnascotia committed:
Fixes when syncing

pugnascotia committed:
Docs tweak

pugnascotia commented on modules/build.gradle:
```diff
@@ -23,6 +23,11 @@ configure(subprojects.findAll { it.parent.path == project.path }) {
   }
   if (project.file('src/main/config').exists()) {
-    throw new InvalidModelException("Modules cannot contain config files")
+    // The `config` directory is different, because we merge together log4j config files
```
@mark-vieira I don't know if this is the right thing to do here. The existing logic in `distribution/build.gradle` will pick up and combine any log4j config files in modules, yet we've banned all config files in modules. I could have just added the relevant logging config to the main config, but I prefer keeping the s3 config close to the relevant source code.

mark-vieira replied:
I think that is all OBE. You can see we don't enforce the same checks for x-pack/plugin (which are modules) and many of them contain config and bin files as well. I don't think we care about having different rules here for x-pack vs non-xpack. And in any case, if these rules were still applicable they should live in `InternalPluginBuildPlugin` not in the project build script.

In any case I think we can remove this. It makes no sense to forbid having config files, especially when we have [explicit handling for this](https://github.com/elastic/elasticsearch/blob/9a3422e1a6cf519e3fedce396784be2ef48dc7f9/distribution/build.gradle#L172-L177).

pugnascotia committed:
Fix GCS largeBlobYamlRestTest

pugnascotia committed:
Attempt to silence lintian issue

pugnascotia committed:
Remove ban on config files in modules

pugnascotia committed:
Fix test

pugnascotia committed:
Merge remote-tracking branch 'upstream/master' into repository-plugins-to-modules

pugnascotia committed:
More docs updates

pugnascotia committed:
Another gcs fix

pugnascotia committed:
Another docs fix

elasticmachine commented:
Pinging @elastic/es-docs (Team:Docs)

pugnascotia committed:
Various lintian changes

Make a number of tweaks to satisfy `lintian`, the `.deb` checker
utility. Of particular note is that `License` is apparently not a valid
field in the Debian control file, so I've removed it. I tweaked a number
of various small things that the `--pedantic` checks reported. I tested
the changes on Debian 9, 10 and 11, with the latest `lintian` available
on each.

pugnascotia committed:
Add migration notes

pugnascotia committed:
Test fix

pugnascotia committed:
Re-add 'License' to deb control file

pugnascotia committed:
Add lintian override for Debian 11

debadair commented on docs/reference/migration/migrate_8_0/plugin-changes.asciidoc:
```diff
@@ -0,0 +1,48 @@
+will succeed but print a warnings.
+
+If you run {es} using Docker and you are managing plugins using a
+<<manage-plugins-using-configuration-file,configuration file>>, then when
```
```suggestion
{plugins}/manage-plugins-using-configuration-file.html[configuration file], then when
```

debadair committed:
Update docs/reference/migration/migrate_8_0/plugin-changes.asciidoc

debadair commented on docs/reference/snapshot-restore/apis/get-repo-api.asciidoc:
```diff
@@ -97,13 +97,11 @@ URL repository. See <<snapshots-read-only-repository>>.
 More repository types are available through these official plugins:
 
-* {plugins}/repository-s3.html[repository-s3] for S3 repository support
+* <<repository-s3,repository-s3>> for S3 repository support
```
```suggestion
* {plugins}/repository-s3[repository-s3] for S3 repository support
```

debadair committed:
Update docs/reference/snapshot-restore/apis/get-repo-api.asciidoc

debadair committed:
[DOCS] Removed reference to installing s3 plugin

debadair committed:
[DOCS] Fixed broken links.

debadair commented:
@elasticmachine retest this please

debadair commented on docs/reference/snapshot-restore/apis/get-repo-api.asciidoc:
```diff
@@ -100,10 +100,8 @@ More repository types are available through these official plugins:
 * {plugins}/repository-s3.html[repository-s3] for S3 repository support
```
```suggestion
* <<repository-s3., repository-s3>> for S3 repository support
```

debadair committed:
Update docs/reference/snapshot-restore/apis/get-repo-api.asciidoc

debadair commented:
@elasticmachine retest this please

debadair commented on docs/reference/snapshot-restore/apis/get-repo-api.asciidoc:
```diff
@@ -97,13 +97,11 @@ URL repository. See <<snapshots-read-only-repository>>.
 More repository types are available through these official plugins:
 
-* {plugins}/repository-s3.html[repository-s3] for S3 repository support
+* <<repository-s3., repository-s3>> for S3 repository support
```
```suggestion
* <<repository-s3, repository-s3>> for S3 repository support
```

debadair committed:
Update docs/reference/snapshot-restore/apis/get-repo-api.asciidoc

debadair commented:
@elasticmachine retest this please

debadair commented:
The last broken links from the Kibana link service are fixed by elastic/kibana#121834. Once this PR is ready to go except for the broken links, it can be merged with the failing doc check and the Kibana PR can be retested & merged.

pugnascotia commented:
@elasticmachine run elasticsearch-ci/cloud-deploy

pugnascotia commented:
This PR requires changes from Cloud before we can merge.

pugnascotia committed:
Merge remote-tracking branch 'upstream/master' into repository-plugins-to-modules

pugnascotia committed:
Merge remote-tracking branch 'upstream/master' into repository-plugins-to-modules

pugnascotia committed:
Merge remote-tracking branch 'origin/repository-plugins-to-modules' into repository-plugins-to-modules
